receiver/mongodbreceiver/client.go

```diff
@@ -0,0 +1,120 @@
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.uber.org/zap"
+)
+
+// client is an interface that exposes functionality towards a mongo environment
+type client interface {
+	ListDatabaseNames(ctx context.Context, filters interface{}, opts ...*options.ListDatabasesOptions) ([]string, error)
+	Disconnect(context.Context) error
+	Connect(context.Context) error
+	GetVersion(context.Context) (*getVersionResponse, error)
+	ServerStatus(ctx context.Context, DBName string) (bson.M, error)
+	DBStats(ctx context.Context, DBName string) (bson.M, error)
+}
+
+// mongodbClient is a mongodb metric scraper client
+type mongodbClient struct {
+	cfg    *Config
+	logger *zap.Logger
+	*mongo.Client
+}
+
+// NewClient creates a new client to connect and query mongo for the
+// mongodbreceiver
+func NewClient(config *Config, logger *zap.Logger) (client, error) {
+	return &mongodbClient{
+		cfg:    config,
+		logger: logger,
+	}, nil
+}
+
+// Connect establishes a connection to mongodb instance
+func (c *mongodbClient) Connect(ctx context.Context) error {
+	if c.Client != nil {
+		return nil
+	}
+
+	driver, err := mongo.Connect(ctx, c.cfg.ClientOptions())
+	if err != nil {
+		return fmt.Errorf("error creating mongo client: %w", err)
+	}
+
+	if err := driver.Ping(ctx, readpref.PrimaryPreferred()); err != nil {
+		return fmt.Errorf("could not connect to hosts %v: %w", c.cfg.hostlist(), err)
+	}
+
+	c.logger.Info(fmt.Sprintf("Mongo connection established to hosts: %s", strings.Join(c.cfg.hostlist(), ", ")))
+	c.Client = driver
+
+	return nil
+}
+
+// RunCommand executes a query against a database. Relies on connection to be established via `Connect()`
+func (c *mongodbClient) RunCommand(ctx context.Context, database string, command bson.M) (bson.M, error) {
+	db := c.Database(database)
+	result := db.RunCommand(ctx, command)
+
+	var document bson.M
+	err := result.Decode(&document)
+	return document, err
+}
+
+// ServerStatus returns the result of db.runCommand({ serverStatus: 1 })
+// more information can be found here: https://docs.mongodb.com/manual/reference/command/serverStatus/
+func (c *mongodbClient) ServerStatus(ctx context.Context, database string) (bson.M, error) {
+	return c.RunCommand(ctx, database, bson.M{"serverStatus": 1})
+}
+
+// DBStats returns the result of db.runCommand({ dbStats: 1 })
+// more information can be found here: https://docs.mongodb.com/manual/reference/command/dbStats/
+func (c *mongodbClient) DBStats(ctx context.Context, database string) (bson.M, error) {
+	return c.RunCommand(ctx, database, bson.M{"dbStats": 1})
+}
+
+type getVersionResponse struct {
+	Version string
+}
```

Comment:
What's the purpose of this?

Reply:
The plan was to use this information to dictate how to collect metrics like lock metrics that have changed from 2.6 to 5.0 like some lock related/switches over to WiredTiger engine metrics so I was going to use this to dictate where metrics are rather than brute force checking all possible metric locations in the documents that we get back from mongo.

Reply:
Ok, but why does it need to be a struct? The value you care about is a string? The struct appears to be entirely unnecessary. Maybe I'm missing something though.

Reply:
Ah no I think I am just returning an overly complex object for a response which I just didn't think too much about. You're right we should just care about the string.